### PR TITLE
Alert fixes without DB migration

### LINKED ISSUE #296: premium account snackbar

### PR DESCRIPTION
**Pull request by milesAraya**
## Summary

This PR fixes 40+ edge cases related to alerts, session management, cross-tab coordination, premium assignment, and frontend resilience. No database migration is required -- all changes are code-level fixes to existing behavior.

This is PR 2 of a 2-PR split from the original `fix/alerts` branch. PR #289 (`fix/critical-with-migration`) contains fixes that require DB schema changes. Both PRs are independent and can be merged in either order.

See fo
[ALERTS_NO_MIGRATION_BRANCH.md](https://github.com/user-attachments/files/25179063/ALERTS_NO_MIGRATION_BRANCH.md)
r more details:

---

### HIGH Priority

| Case | Description |
|------|-------------|
| 3 | Browser closed during logout -- premium instance not released |
| 9 | getPremiumStatus fails during logout -- instance leaked |
| 20 | S3 deletion succeeds but local FS deletion fails |
| 24 | Cleanup job vs new workflow race -- data partially deleted |
| 33 | User dismisses OVERDUE alert and ignores -- no forced action |
| 39 | Pre-flight storage check fails but job proceeds anyway |
| 49 | Heartbeat fails silently -- false inactivity detection |
| 52 | Upload in Tab 1, delete workspace in Tab 2 -- race condition |
| 58/62 | Free user logged_out_at never cleared on re-login |
| 68 | API timeout during premium instance assignment |
| 73 | Payment fails during renewal -- silent failure |
| 76 | User tier cache not invalidated on subscription change |
| 82 | No React error boundary -- white screen on crash |
| 90 | Database datetime stored as naive -- comparison drift |
| 96 | Free user shown "Premium Expired" alert |
| 97 | New user has no storage quota record |

### MEDIUM Priority

| Case | Description |
|------|-------------|
| 1 | Rate-limited after rapid logout/login within 30s |
| 4 | Multiple tabs poll independently -- no coordination |
| 10 | Activity cache stale after rapid re-login |
| 11 | Background activity update runs after logout |
| 13 | Stale token in React closures after logout |
| 28 | Multiple alert types display simultaneously -- no priority |
| 30 | Alert API error during polling -- silent failure |
| 31 | Alert dismissal not synced across tabs |
| 36 | Premium storage issue misidentified as free user |
| 42 | Storage refresh timeout -- stale data persists |
| 43 | Polling backoff for premium assignment not implemented |
| 51 | Computer sleep triggers false inactivity detection |
| 53 | Job running in Tab 1, logout in Tab 2 |
| 77 | Trial-to-paid conversion with 8+ day gap missed |
| 79 | Storage quota instantly drops on downgrade |
| 81 | Browser tab hidden but polling continues |
| 83 | localStorage quota exceeded crashes dismissal |
| 84 | Modal z-index collisions |
| 89 | "0 days remaining" when ~24 hours left |
| 95 | Premium idle timeout vs activity tracking mismatch |

### LOW Priority

| Case | Description |
|------|-------------|
| 5 | Session expires during warning countdown |
| 8 | Redux logout dispatch while components still mounted |
| 12 | Inactivity interval fires during logout |
| 29 | Duplicate snackbar alerts |
| 32 | Dismissed alerts persist in localStorage after logout |
| 34 | StorageAlert refetched after dismiss |
| 35 | Negative days remaining value |
| 37 | Grace period expires while user offline |
| 38 | Subscription renews during grace period -- alert persists |
| 50 | Dismiss inactivity warning but auto-release proceeds |
| 54-56 | Cross-tab state staleness |
| 59 | Storage alert persists in localStorage after logout |

---

### Frontend - Core Application (3 files)

| File | Change | Cases |
|------|--------|-------|
| `App.tsx` | ErrorBoundary wrapper, preventDuplicate snackbar | 29, 82 |
| `const/Layout.ts` | Centralized Z_INDEX constants | 84 |
| `const/Subscription.ts` | PremiumTiming constants for frontend/backend sync | 95 |

### Frontend - Components (9 files)

| File | Change | Cases |
|------|--------|-------|
| `common/ErrorBoundary.tsx` | NEW: React error boundary with retry | 82 |
| `common/LimitAlert.tsx` | Cross-tab dismissal, OVERDUE acknowledgment, negative days fix, time formatting | 31, 33, 35, 89 |
| `common/Loading.tsx` | Centralized Z_INDEX constant | 84 |
| `common/LogsFloatingButton.tsx` | Centralized Z_INDEX constant | 84 |
| `PublicLayout/PublicHeader.tsx` | Centralized Z_INDEX constant | 84 |
| `Layout/Profile.tsx` | Running job check before logout, cross-tab logout sync | 53, 54-56 |
| `Layout/index.tsx` | Storage refresh timeout and retry logic | 42 |
| `Workspace/FlowChart/Buttons/RunButtons.tsx` | Storage check confirmation dialog | 39 |
| `Premium/InactivityWarning.tsx` | Inactivity warning improvements | 50 |

### Frontend - Contexts (2 files)

| File | Change | Cases |
|------|--------|-------|
| `contexts/PremiumAssignmentContext.tsx` | Sleep detection, heartbeat retry, polling backoff, logout guard, sendBeacon cleanup | 3, 9, 12, 13, 43, 49, 50-51 |
| `contexts/AlertPriorityContext.tsx` | NEW: Alert prioritization system | 28 |

### Frontend - Hooks (3 files)

| File | Change | Cases |
|------|--------|-------|
| `hooks/useVisibilityAwarePolling.ts` | NEW: Pause polling when tab hidden | 81 |
| `hooks/useSleepDetection.ts` | NEW: Detect device wake from sleep | 50-51 |
| `hooks/useLimitAlert.ts` | Error tracking, retry logic, visibility change detection | 30, 36, 37, 38 |

### Frontend - Utils (4 files)

| File | Change | Cases |
|------|--------|-------|
| `utils/safeStorage.ts` | NEW: Safe localStorage wrapper with quota handling | 83 |
| `utils/operationLock.ts` | NEW: Cross-tab workspace operation locking | 52 |
| `utils/crossTabSync.ts` | NEW: Leader election, TabSyncService for cross-tab communication | 4, 54-56 |
| `utils/axios.ts` | Debounced 401 handling, logout flag timing | 6, 8 |

### Frontend - Store (5 files)

| File | Change | Cases |
|------|--------|-------|
| `store/slice/User/UserSlice.ts` | logoutGeneration counter, clear dismissed alerts on logout | 8, 32, 59 |
| `store/slice/User/UserType.ts` | logoutGeneration field | 8 |
| `store/slice/User/UserSelector.ts` | selectLogoutGeneration selector | 8 |
| `store/slice/FileUploader/FileUploaderHook.ts` | Acquire/release operation lock during uploads | 52 |
| `store/slice/Pipeline/PipelineHook.ts` | Acquire/release operation lock during workflow runs | 52 |

### Frontend - API (1 file)

| File | Change | Cases |
|------|--------|-------|
| `api/workspace/index.ts` | AbortController signal support for storage refresh | 42 |

### Backend - Core Services (8 files)

| File | Change | Cases |
|------|--------|-------|
| `core/background/cleanup_job.py` | Re-login detection before cleanup | 24 |
| `core/experiment/experiment_writer.py` | S3 deletion retry with backoff, handle local deletion separately | 15, 20 |
| `core/premium/premium_assignment_service.py` | Rate limit persistence, Lambda timeout handling | 1, 68 |
| `core/subscription/constants.py` | Extended lookback window, quota drop warning days | 77, 79 |
| `core/subscription/subscription_service.py` | get_users_with_upcoming_quota_drop() | 79 |
| `core/subscription/webhook_service.py` | Payment failure tracking, cache invalidation | 73, 76, 77 |
| `core/utils/datetime_utils.py` | ensure_utc(), is_datetime_aware() functions | 90 |
| `routers/auth.py` | Clear logged_out_at on login | 58/62 |

### Backend - Shared Files (surgical split, 3 files)

These files are shared with PR #289. Each PR contains only its own changes applied to the `develop-subscription` base.

| File | PR 2 Changes | Cases |
|------|-------------|-------|
| `core/cloud/cloud_utils.py` | Premium-only filter in calculate_limit_warning, non-exceeded case | 96 |
| `core/middleware/user_activity_middleware.py` | Activity cache invalidation, logged-out tracking, clear_free_user_logged_out_at | 10, 11, 58/62 |
| `routers/users_me.py` | /premium/release-beacon endpoint, cache invalidation on logout | 3, 10, 11 |

### Backend - Tests (16 files)

| File | Tests For | Cases |
|------|-----------|-------|
| `test_cleanup_job.py` | Cleanup job edge cases | 24 |
| `test_cleanup_job_relogin.py` | Re-login detection | 24 |
| `test_cloud_utils.py` | Premium filter, storage freshness, limit warnings | 96 |
| `test_experiment_deletion_recovery.py` | S3 deletion retry with backoff | 15, 20 |
| `test_user_activity_middleware.py` | Activity cache, logged-out tracking, cross-tab cleanup | 10, 11, 58/62 |
| `test_rate_limit_persistence.py` | Rate limit persistence, Lambda timeout | 1, 68 |
| `test_remote_storage_controller.py` | Sync status on partial failure | 69 |
| `test_subscription_service.py` | Quota drop warning | 79 |
| `test_storage_tracking.py` | Storage tracking edge cases | 97 |
| `test_datetime_utils.py` | ensure_utc, is_datetime_aware | 90 |
| `test_premium_api_contract.py` | Premium API contract tests | 3 |
| `test_premium_beacon_endpoint.py` | Beacon endpoint unit tests | 3 |
| `test_data_sync.py` | Data sync edge cases | -- |
| `test_auth_api_contract.py` | Login clears logged_out_at | 58/62 |
| `test_storage_limit_alerts.py` | Storage limit alert API tests | 96 |
| `test_webhook.py` | Payment failure tracking, cache invalidation | 73, 76, 77 |

### Frontend - Tests (16 files)

| File | Tests For | Cases |
|------|-----------|-------|
| `ErrorBoundary.test.tsx` | Error boundary behavior | 82 |
| `LimitAlert.test.tsx` | Cross-tab dismissal, negative days, time formatting | 31, 33, 35, 89 |
| `ExperimentTable.test.tsx` | Experiment table edge cases | -- |
| `RunButtons.test.tsx` | Storage check confirmation dialog | 39 |
| `AlertPriorityContext.test.tsx` | Alert priority system | 28 |
| `PremiumHeartbeatRetry.test.ts` | Heartbeat retry logic | 49 |
| `PremiumPollingBackoff.test.ts` | Polling backoff behavior | 43 |
| `PremiumSleepDetection.test.ts` | Sleep detection integration | 50-51 |
| `useLimitAlert.test.ts` | Limit alert hook | 30, 36-38 |
| `useSleepDetection.test.ts` | Sleep detection hook | 50-51 |
| `useVisibilityAwarePolling.test.ts` | Visibility-aware polling | 81 |
| `UserSlice.test.ts` | logoutGeneration, clear dismissed alerts | 8, 32, 59 |
| `axios.test.ts` | Debounced 401 handling | 6 |
| `crossTabSync.test.ts` | Leader election, TabSyncService | 4, 54-56 |
| `operationLock.test.ts` | Cross-tab operation locking | 52 |
| `safeStorage.test.ts` | Safe localStorage wrapper | 83 |
| `workspace.test.ts` | Workspace API with AbortController | 42 |

---

## Test Results

```
Backend:  271 passed, 0 failed
Frontend: 393 passed, 5 skipped (38 suites all pass)
```

---

## Risk Assessment

| Area | Risk | Mitigation |
|------|------|------------|
| Cross-tab locking | localStorage race conditions | 5-minute stale lock detection |
| Sleep detection | False positives on wake | 30s threshold, only triggers refresh |
| sendBeacon cleanup | Browser may drop beacon | Stale assignment cleanup Lambda as fallback |
| Leader election | Split-brain across tabs | Heartbeat-based leader with 10s TTL |
| Error boundary | Swallows useful errors | Logs error details, shows retry button |

---

## Relationship to PR #289

PR #289 (`fix/critical-with-migration`) contains 6 cases requiring DB migration:
- Case 25 (CRITICAL): Two-phase Firebase deletion
- Cases 14, 18: Deletion task queue and orphan tracking
- Case 23: Workspace row-level locking
- Case 69: Idempotent storage operations
- Case 71: Heartbeat failure tracking

3 files (`cloud_utils.py`, `user_activity_middleware.py`, `users_me.py`) are surgically split so each PR contains only its own changes. Both PRs can be merged independently in any order.

### CODE REVIEW

File: frontend/src/App.tsx [hunk 1]
```diff
@@ -7,6 +7,7 @@ import { SnackbarProvider, SnackbarKey, useSnackbar } from "notistack"
 import Close from "@mui/icons-material/Close"
 import IconButton from "@mui/material/IconButton"
 
+import ErrorBoundary from "components/common/ErrorBoundary"
```
Comment (itutu-tienday) [inline frontend/src/App.tsx:10]: This PR applies quite a large number of fixes, but for the purpose of review (and future maintenance), would it be okay if you could compile a final list of Alert Events (classified by severity, category, etc.) into a document?
We will use that document as a reference when conducting our review.

*The text "ALERT_FIX_SUMMARY.md" is attached, but this appears to be a difference in the update content, so we would like to see a summary of the entire document.

Comment (milesAraya) [reply frontend/src/App.tsx:10]: See [ALERTS_NO_MIGRATION_BRANCH.md](https://github.com/user-attachments/files/25179063/ALERTS_NO_MIGRATION_BRANCH.md)

**Review by itutu-tienday — COMMENTED** [review]
@milesAraya 
Please proceed to deploy to check that it works.
Approval will be withheld until the operation has been checked.

**Review by itutu-tienday — APPROVED** [review]
This content has already been deployed and tested in the verification environment, so I will mark it as approved.
